File: patch_db.js
const fs = require('fs');
let code = fs.readFileSync('src/server.ts', 'utf-8');
code = code.replace(
  '  try {\n    db.exec("ALTER TABLE invoices ADD COLUMN services_json TEXT");',
  '  try {\n    db.exec("ALTER TABLE shifts ADD COLUMN is_historical INTEGER DEFAULT 0");\n    console.log("[DEBUG] Completed is_historical column check.");\n  } catch (e: any) {\n    if (e.message && !e.message.includes("duplicate column")) {\n      console.warn("Migration warning:", e.message);\n    }\n  }\n\n  try {\n    db.exec("ALTER TABLE invoices ADD COLUMN services_json TEXT");'
);
fs.writeFileSync('src/server.ts', code);

File: patch_db.py
import re

with open("src/server.ts", "r") as f:
    code = f.read()

migration_code = """
  try {
    const shiftCols = db.pragma("table_info(shifts)") as any[];
    const colNames = shiftCols.map((c: any) => c.name);
    const colsToAdd = [
      ["transport_route_log", "TEXT"],
      ["services_json", "TEXT"],
      ["home_care_travel_km", "REAL DEFAULT 0"],
      ["home_care_travel_total", "REAL DEFAULT 0"],
      ["batch_id", "TEXT"],
      ["odometer_start_photo", "TEXT"],
      ["odometer_end_photo", "TEXT"],
      ["odometer_start_reading", "REAL"],
      ["odometer_end_reading", "REAL"],
      ["merged_into_shift_id", "INTEGER"]
    ];
    for (const [colName, colType] of colsToAdd) {
      if (!colNames.includes(colName)) {
        db.exec(`ALTER TABLE shifts ADD COLUMN ${colName} ${colType}`);
        console.log(`[DEBUG] Added missing column shifts.${colName}`);
      }
    }
  } catch (e: any) {
    console.warn("Migration warning for shifts cols:", e.message);
  }

  try {
    db.exec(`
      CREATE TABLE IF NOT EXISTS settings (
        key TEXT PRIMARY KEY,
        value TEXT NOT NULL
      );
    `);
    console.log("[DEBUG] Created settings table if missing");
  } catch (e: any) {
    console.warn("Migration warning for settings table:", e.message);
  }
"""

code = code.replace('db.exec("ALTER TABLE shifts ADD COLUMN is_historical INTEGER DEFAULT 0");', 'db.exec("ALTER TABLE shifts ADD COLUMN is_historical INTEGER DEFAULT 0");\n' + migration_code)

with open("src/server.ts", "w") as f:
    f.write(code)
